### PR TITLE
test: Add more tests for orphan tx handling

### PR DESCRIPTION
**Pull request by hebasto**
This PR adds test coverage for the following cases:
- erase orphan transactions when a peer is disconnected
- erase an orphan transaction when it is included in a new tip block
- erase an orphan transaction when it is conflicted with other transactions included in a new tip block

Found useful while working on #19374.

### CODE REVIEW

File: test/functional/p2p_invalid_tx.py [hunk 1]
```diff
@@ -158,13 +157,71 @@ def run_test(self):
...
         rejected_parent.vout.append(CTxOut(nValue=11 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
         rejected_parent.rehash()
         with node.assert_debug_log(['not keeping orphan with rejected parents {}'.format(rejected_parent.hash)]):
             node.p2p.send_txs_and_test([rejected_parent], node, success=False)
 
+        self.log.info('Test erase of orphan tx from peer')
+        with node.assert_debug_log(['Erased 100 orphan tx from peer=24']):
+            self.reconnect_p2p(num_connections=1)
+
+        self.log.info('Test erase of orphan tx included by block')
```
Comment (jonatack) [inline test/functional/p2p_invalid_tx.py:172]: What is being asserted in this subtest?

File: test/functional/p2p_invalid_tx.py [hunk 2]
```diff
@@ -158,13 +157,71 @@ def run_test(self):
...
+        tx_withhold_until_blockA.vout = [CTxOut(nValue=12 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 2
+        tx_withhold_until_blockA.calc_sha256()
+
+        tx_orphan_include_by_blockA = CTransaction()
+        tx_orphan_include_by_blockA.vin.append(CTxIn(outpoint=COutPoint(tx_withhold_until_blockA.sha256, 0)))
+        tx_orphan_include_by_blockA.vout.append(CTxOut(nValue=12 * COIN - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
+        tx_orphan_include_by_blockA.calc_sha256()
+
+        self.log.info('Send the tx_orphan_include_by_blockA ... ')
+        node.p2p.send_txs_and_test([tx_orphan_include_by_blockA], node, success=False)
```
Comment (jonatack) [inline test/functional/p2p_invalid_tx.py:184]: This is checking the the txn is not accepted into the mempool; mind mentioning the expected outcome in the info log and why? ISTM this is the assertion for the subtest beginning at line 172? Perhaps just one log at 172 that describes the assertion/expected outcome/why, then one at the start of the next subtest that does the same.

File: test/functional/p2p_invalid_tx.py [hunk 3]
```diff
@@ -158,13 +157,71 @@ def run_test(self):
...
+
+        tip = int(node.getbestblockhash(), 16)
+        height = node.getblockcount() + 1
+        blockA = create_block(tip, create_coinbase(height))
+        blockA.vtx.extend([tx_withhold, tx_withhold_until_blockA, tx_orphan_include_by_blockA])
+        blockA.hashMerkleRoot = blockA.calc_merkle_root()
+        blockA.solve()
+
+        self.log.info('Send the blockA that includes tx_orphan_include_by_blockA ... ')
```
Comment (jonatack) [inline test/functional/p2p_invalid_tx.py:193]: Idem, mind describing the expected outcome (accepted by the mempool) and why? This log seems more cryptic than reading the test code.

File: test/functional/p2p_invalid_tx.py [hunk 4]
```diff
@@ -158,13 +157,71 @@ def run_test(self):
...
+        self.log.info('Test orphan with rejected parents')
         rejected_parent = CTransaction()
         rejected_parent.vin.append(CTxIn(outpoint=COutPoint(tx_orphan_2_invalid.sha256, 0)))
         rejected_parent.vout.append(CTxOut(nValue=11 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
         rejected_parent.rehash()
         with node.assert_debug_log(['not keeping orphan with rejected parents {}'.format(rejected_parent.hash)]):
             node.p2p.send_txs_and_test([rejected_parent], node, success=False)
 
+        self.log.info('Test erase of orphan tx from peer')
+        with node.assert_debug_log(['Erased 100 orphan tx from peer=24']):
```
Comment (jonatack) [inline test/functional/p2p_invalid_tx.py:169]: This test seems dependent on the context/test order. Is it possible to make it order-independent and to test for the change itself?

**Review by jonatack — COMMENTED** [review]
Concept ACK.

- Could say in the PR description what led you to see the need for this additional coverage.

- Testing debug log output is convenient but less robust; better to test for actual change, if possible.

- Readability nit: for variable names when the rest of the name is snakecased: `s/blockA/block_A` and `s/blockB/block_B/`

- It would be great for the logging to state clearly what the test is asserting, the expected outcome, and why. ISTM `info` logging can be good at the beginning of a subtest as well as just before the key assertion(s).

- If a subtest does not need the context from the previous ones, it can be good to separate it to an independent method.

**Comment by hebasto** [comment]
@jonatack 
> Could say in the PR description what led you to see the need for this additional coverage.

The OP has been updated.

> Testing debug log output is convenient but less robust; better to test for actual change, if possible.

Agree. But we have neither RPC calls nor the public interface for the orphan transaction pool.

File: test/functional/p2p_invalid_tx.py [hunk 5]
```diff
@@ -94,24 +93,24 @@ def run_test(self):
         SCRIPT_PUB_KEY_OP_TRUE = b'\x51\x75' * 15 + b'\x51'
         tx_withhold = CTransaction()
         tx_withhold.vin.append(CTxIn(outpoint=COutPoint(block1.vtx[0].sha256, 0)))
-        tx_withhold.vout.append(CTxOut(nValue=50 * COIN - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
+        tx_withhold.vout = [CTxOut(nValue=25 * COIN  - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 2
```
[the remark this answers is not in the document]
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py:96]: hmm, why python linter did not catch it?

File: test/functional/p2p_invalid_tx.py (hunk 1 above)
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py:172]: > What is being asserted in this subtest?

If a transaction from a new tip block is found in the orphan pool, it should be excluded from it.

File: test/functional/p2p_invalid_tx.py (hunk 4 above)
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py:169]: This variant, i.e., the previous test state reuse, just suggests the minimum diff.

**Comment by hebasto** [comment]
Updated 10f7f75502bdcc9045c0623386571e41103b69fe -> a95a6d99b31798a8adb107930857bbe5c6e083df ([pr19393.01](https://github.com/hebasto/bitcoin/commits/pr19393.01) -> [pr19393.02](https://github.com/hebasto/bitcoin/commits/pr19393.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19393.01..pr19393.02)):

- addressed @jonatack's comments

File: test/functional/p2p_invalid_tx.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py]: [Updated](https://github.com/bitcoin/bitcoin/pull/19393#issuecomment-650795757).

File: test/functional/p2p_invalid_tx.py [hunk 6]
```diff
@@ -158,13 +157,71 @@ def run_test(self):
...
 
+        self.log.info('Test orphan with rejected parents')
         rejected_parent = CTransaction()
         rejected_parent.vin.append(CTxIn(outpoint=COutPoint(tx_orphan_2_invalid.sha256, 0)))
         rejected_parent.vout.append(CTxOut(nValue=11 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
         rejected_parent.rehash()
         with node.assert_debug_log(['not keeping orphan with rejected parents {}'.format(rejected_parent.hash)]):
             node.p2p.send_txs_and_test([rejected_parent], node, success=False)
 
+        self.log.info('Test erase of orphan tx from peer')
```
[the remark this answers is not in the document]
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py:168]: I'd rather leave log message as is, as working logging implies working erase a tx from the orphan pool.

File: test/functional/p2p_invalid_tx.py (hunk 2 above)
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py]: Re-worded. It seems mention "orphan" transaction implies that it is not accepted into the mempool, no?

File: test/functional/p2p_invalid_tx.py (hunk 3 above)
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py]: [Re-worded](https://github.com/bitcoin/bitcoin/pull/19393#issuecomment-650795757).

File: test/functional/p2p_invalid_tx.py (hunk 1 above)
Comment (jonatack) [reply test/functional/p2p_invalid_tx.py:172]: Could add that info to the log message. Same with https://github.com/bitcoin/bitcoin/pull/19393#discussion_r446675398: state the action and expected result under test: "Test <action> causes <result>". Just my opinion, of course.

**Comment by hebasto** [comment]
Updated a95a6d99b31798a8adb107930857bbe5c6e083df -> 3a310fbf1733e2ddaf1e6707251c409b287838de ([pr19393.02](https://github.com/hebasto/bitcoin/commits/pr19393.02) -> [pr19393.03](https://github.com/hebasto/bitcoin/commits/pr19393.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19393.02..pr19393.03)):

- addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/19393#discussion_r446676922):
> Could add that info to the log message. Same with [#19393 (comment)](https://github.com/bitcoin/bitcoin/pull/19393#discussion_r446675398): state the action and expected result under test: "Test (action) causes (result)". Just my opinion, of course.

File: test/functional/p2p_invalid_tx.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py]: [Updated](https://github.com/bitcoin/bitcoin/pull/19393#issuecomment-650802681).

File: test/functional/p2p_invalid_tx.py (hunk 1 above)
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py]: [Updated](https://github.com/bitcoin/bitcoin/pull/19393#issuecomment-650802681).

**Comment by hebasto** [comment]
Rebased 3a310fbf1733e2ddaf1e6707251c409b287838de -> 9979e8b96f475762c9ec4feb98bcd80c029b1913 ([pr19393.03](https://github.com/hebasto/bitcoin/commits/pr19393.03) -> [pr19393.04](https://github.com/hebasto/bitcoin/commits/pr19393.04)) due to the conflict with #19804.

**Comment by hebasto** [comment]
Rebased 3a310fbf1733e2ddaf1e6707251c409b287838de -> 8e6c42797e0c5b30df34f464eb3b1cddc477da26 ([pr19393.03](https://github.com/hebasto/bitcoin/commits/pr19393.03) -> [pr19393.05](https://github.com/hebasto/bitcoin/commits/pr19393.05)) due to the conflict with #19804.

**Comment by Emzy** [comment]
tACK
Tested and worked:
src/test/fuzz/addrman
src/test/fuzz/kitchen_sink
src/test/fuzz/merkleblock

**Comment by NelsonGaldeman** [comment]
Concept ACK

- I have tested the code

**Comment by leonardojobim** [comment]
Concept ACK. It expands the test cases that cause the removal of the orphan transactions.

I also agree it would be better to test against the actual changes, but as far I know there's no way in these tests to access  the orphan pool maintained in the `mapOrphanTransactionsdata` structure to perform this validation.

I ran successfully this test on Ubuntu 20.04.

**Review by leonardojobim — APPROVED** [review]
Concept ACK. It expands the test cases that cause the removal of the orphan transactions.

I also agree it would be better to test against the actual changes, but as far I know there's no way in these tests to access the orphan pool maintained in the `mapOrphanTransactionsdata` structure to perform this validation.

I ran successfully this test on Ubuntu 20.04.

**Review by aarmoa — APPROVED** [review]
Approach ACK.
To validate the new verification work as expected I did the following temporary change while running the test locally:
Line 168:
- If assertion is done for 99 number of orphans instead of 100, test fails.
- If assertion is done for node 23 instead of 24, test fails.

Line 190:
- Added a node.assert_debug_log() with the message "Erased 1 orphan tx included or conflicted by block" as unexpected, to ensure it is logged only when block_A is sent. The test runs without failures.

Line 219:
- Added a node.assert_debug_log() with the message "Erased 1 orphan tx included or conflicted by block" as unexpected, to ensure it is logged only when block_B is sent. The test runs without failures.

Tested in Linux Mint 20 64 bits.

**Comment by practicalswift** [comment]
Concept ACK

File: test/functional/p2p_invalid_tx.py [hunk 7]
```diff
@@ -93,24 +92,24 @@ def run_test(self):
         SCRIPT_PUB_KEY_OP_TRUE = b'\x51\x75' * 15 + b'\x51'
         tx_withhold = CTransaction()
         tx_withhold.vin.append(CTxIn(outpoint=COutPoint(block1.vtx[0].sha256, 0)))
-        tx_withhold.vout.append(CTxOut(nValue=50 * COIN - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
+        tx_withhold.vout = [CTxOut(nValue=25 * COIN - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 2
```
Comment (ShubhamPalriwala) [inline test/functional/p2p_invalid_tx.py:95]: nit: Having `append()` here would maintain consistency across the entire file

File: test/functional/p2p_invalid_tx.py [hunk 8]
```diff
@@ -93,24 +92,24 @@ def run_test(self):
...
         tx_withhold.calc_sha256()
 
         # Our first orphan tx with some outputs to create further orphan txs
         tx_orphan_1 = CTransaction()
         tx_orphan_1.vin.append(CTxIn(outpoint=COutPoint(tx_withhold.sha256, 0)))
-        tx_orphan_1.vout = [CTxOut(nValue=10 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 3
+        tx_orphan_1.vout = [CTxOut(nValue=8 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 3
```
Comment (ShubhamPalriwala) [inline test/functional/p2p_invalid_tx.py:101]: nit: Having `append()` here would maintain consistency across the entire file

File: test/functional/p2p_invalid_tx.py [hunk 9]
```diff
@@ -157,13 +156,72 @@ def run_test(self):
...
             node.p2ps[0].send_txs_and_test([rejected_parent], node, success=False)
 
+        self.log.info('Test that a peer disconnection causes erase its transactions from the orphan pool')
+        with node.assert_debug_log(['Erased 100 orphan tx from peer=24']):
+            self.reconnect_p2p(num_connections=1)
+
+        self.log.info('Test that a transaction in the orphan pool is included in a new tip block causes erase this transaction from the orphan pool')
+        tx_withhold_until_block_A = CTransaction()
+        tx_withhold_until_block_A.vin.append(CTxIn(outpoint=COutPoint(tx_withhold.sha256, 1)))
+        tx_withhold_until_block_A.vout = [CTxOut(nValue=12 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 2
```
Comment (ShubhamPalriwala) [inline test/functional/p2p_invalid_tx.py:174]: nit: Having `append()` here would maintain consistency across the entire file

**Review by ShubhamPalriwala — COMMENTED** [review]
tACK 73278e3fda86f89ae70b8922f029e6b2d72d6bd6! 

Suggestion: the `log.info()` messages across the PR can be framed in a more clear and concise way.

I completely agree with @leonardojobim here!
> Concept ACK. It expands the test cases that cause the removal of the orphan transactions.
> 
> I also agree it would be better to test against the actual changes, but as far I know there's no way in these tests to access the orphan pool maintained in the `mapOrphanTransactionsdata` structure to perform this validation.
> 
> I ran successfully this test on Ubuntu 20.04.

Tested on Ubuntu 21.04 works fine

**Comment by MarcoFalke** [comment]
> Tests don't pass for me on the first commit.

After a rebase, that is.

**Comment by hebasto** [comment]
> Needs rebase.
> 
> Tests don't pass for me on the first commit.


Rebased 8e6c42797e0c5b30df34f464eb3b1cddc477da26 -> c0a5fceee9858afd24fe0bf655b7b30728e96e78 ([pr19393.05](https://github.com/hebasto/bitcoin/commits/pr19393.05) -> [pr19393.06](https://github.com/hebasto/bitcoin/commits/pr19393.06)).

File: test/functional/p2p_invalid_tx.py (hunk 7 above)
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py:95]: See the current code in the master branch https://github.com/bitcoin/bitcoin/blob/dd097c42df7378c9d0114c04c82b3c38c18e13dc/test/functional/p2p_invalid_tx.py#L102

File: test/functional/p2p_invalid_tx.py (hunk 8 above)
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py:101]: See the current code in the master branch https://github.com/bitcoin/bitcoin/blob/dd097c42df7378c9d0114c04c82b3c38c18e13dc/test/functional/p2p_invalid_tx.py#L102

File: test/functional/p2p_invalid_tx.py (hunk 9 above)
Comment (hebasto) [reply test/functional/p2p_invalid_tx.py:174]: See the current code in the master branch https://github.com/bitcoin/bitcoin/blob/dd097c42df7378c9d0114c04c82b3c38c18e13dc/test/functional/p2p_invalid_tx.py#L102

**Review by pg156 — COMMENTED** [review]
Reviewed to https://github.com/bitcoin/bitcoin/pull/19393/commits/c0a5fceee9858afd24fe0bf655b7b30728e96e78. Concept ACK. Agree due to the lack of RPC calls to inspect orphan pool, using `assert_debug_log` to match strings in log is a reasonable way to test.

Tested successfully by running `make check` and `test/functional/test_runner.py`.

If I understand correctly, https://github.com/bitcoin/bitcoin/pull/19393/commits/5c049780c8b310428cf72fb304bf0c1071742785 tests 100 orphan transactions sent by a peer to the node are removed after the peer is disconnected. (The peer initially sent 101 orphan transactions to the node. 1 orphan tx is removed because the default max orphan pool size `-maxorphantx` is 100.) This makes sense.

In https://github.com/bitcoin/bitcoin/commit/c0a5fceee9858afd24fe0bf655b7b30728e96e78, an orphan transaction is sent first to node, then a block with a conflicting transaction is sent to node, resulting in removing the orphan transaction. This is also reasonable.

Questions and suggestions are in below.

**Comment by aureleoules** [comment]
tACK c0a5fceee9858afd24fe0bf655b7b30728e96e78 (`make check` and `test/functional/test_runner.py`).
Tested on NixOS 22.05 64 bits.  

I commented out locally [`src/txorphanage.cpp:189`](https://github.com/hebasto/bitcoin/blob/200627-test-orphan/src/txorphanage.cpp#L189) and the tests fail accordingly.  
Added tests seem to correctly test the behavior.  

nit: I would agree with [pg156](https://github.com/pg156), magic numbers should be changed to variables for increased readability and better comprehension.  

nit: I found it a bit confusing the use of `*_A` and `*_B` variables as I first thought the two tests were sequentially dependent. Would it make more sense to use unique variables that would get overridden by the second test, or separate both tests in different methods completely? I hope this is not too _nitty_ but it was my initial thought.

File: test/functional/p2p_invalid_tx.py [hunk 10]
```diff
@@ -157,13 +156,72 @@ def run_test(self):
...
+        tx_orphan_include_by_block_A.vout.append(CTxOut(nValue=12 * COIN - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
+        tx_orphan_include_by_block_A.calc_sha256()
+
+        self.log.info('Send the orphan ... ')
+        node.p2ps[0].send_txs_and_test([tx_orphan_include_by_block_A], node, success=False)
+
+        tip = int(node.getbestblockhash(), 16)
+        height = node.getblockcount() + 1
+        block_A = create_block(tip, create_coinbase(height))
+        block_A.vtx.extend([tx_withhold, tx_withhold_until_block_A, tx_orphan_include_by_block_A])
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply test/functional/p2p_invalid_tx.py:188]: When the grandchild was added it was missing inputs thus added to the orphan pool, but those orphans are not reevaluated at each newly connected block. 

Taking a look into the codebase, when a new orphan tx is added, there are a few indexes to the OrhanMap, [one of them is from the parents' COutPoint to the tx](https://github.com/hebasto/bitcoin/blob/c0a5fceee9858afd24fe0bf655b7b30728e96e78/src/txorphanage.cpp#L47-L49). On a newly connected block, [that index is used](https://github.com/hebasto/bitcoin/blob/c0a5fceee9858afd24fe0bf655b7b30728e96e78/src/txorphanage.cpp#L184) to check if the included transaction's parents have connection to any other txns from the orphan pool, and those are evicted. 

So in you example, when child (`tx_withhold_until_block_A`) is evaluated, eviction looks for txns in the pool also connected to the parent (`tx_withhold`) and not grandchildren (`tx_orphan_include_by_block_A`) connected to the child.

**Comment by kouloumos** [comment]
ACK c0a5fceee9858afd24fe0bf655b7b30728e96e78 with a nit per https://github.com/bitcoin/bitcoin/pull/19393#discussion_r899156623.

The description of the flow in https://github.com/bitcoin/bitcoin/pull/19393#discussion_r900233637 might be useful for reviewers to understand the behavior tested on fa45bb21193ae0c220cfc224d5e3ea0e7f3ec988 & c0a5fceee9858afd24fe0bf655b7b30728e96e78.

**Pull request merged**